packages/filecoin-api/src/storefront/service.js

```diff
@@ -0,0 +1,260 @@
+import * as Server from '@ucanto/server'
+import * as Client from '@ucanto/client'
+import * as CAR from '@ucanto/transport/car'
+import * as StorefrontCaps from '@web3-storage/capabilities/filecoin/storefront'
+import * as AggregatorCaps from '@web3-storage/capabilities/filecoin/aggregator'
+// eslint-disable-next-line no-unused-vars
+import * as API from '../types.js'
+import { QueueOperationFailed, StoreOperationFailed } from '../errors.js'
+
+/**
+ * @param {API.Input<StorefrontCaps.filecoinOffer>} input
+ * @param {import('./api').ServiceContext} context
+ * @returns {Promise<API.UcantoInterface.Result<API.FilecoinOfferSuccess, API.FilecoinOfferFailure> | API.UcantoInterface.JoinBuilder<API.FilecoinOfferSuccess>>}
+ */
+export const filecoinOffer = async ({ capability }, context) => {
+  const { piece, content } = capability.nb
+
+  // Queue offer for filecoin submission
+  if (!context.config?.skipFilecoinSubmitQueue) {
+    // dedupe
+    const hasRes = await context.pieceStore.has({ piece })
+    if (hasRes.error) {
+      return { error: new StoreOperationFailed(hasRes.error.message) }
+    }
+    const exists = hasRes.ok
```

Comment:
Elsewhere I think we're using result error to communicate not found - I think I'm currently doing the same in billing. Can we come to an accord?